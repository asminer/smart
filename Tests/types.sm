


print("\nThese should all have type void:\n\n");

print_type(print(null));
print_type(print("foo\n"));
print_type(cond(true, print("bar\n"), null));

print("\nThese should have type int:\n\n");

int a := 3;
int b := 4;
print_type(2);
print_type(a);
print_type(2+a);
print_type(b-a);
print_type(a*b);

print("\nThese should have type real:\n\n");

real c := 4.2;
print_type(5.1);
print_type(5.1 + a);
print_type(b+c);
print_type(b-c);
print_type(b*c);
print_type(b/c);
print_type(a/b);

// Will want to add ph and rand tests here...

print("\nAggregate types:\n\n");

print_type(a:b:c);
print_type(3:4:null);

print("\nTesting function promotions:\n\n");

int myfunc(int n) := n*n+1;
real myfunc(real x) := sqrt(x+1);
rand int R := equilikely(1, 6);
rand real X := uniform(1.0, 6.0);

print("Should be int:\n");
print_type(myfunc(3));
print("Should be rand int:\n");
print_type(myfunc(R));
print("Should be real:\n");
print_type(myfunc(3.5));
print("Should be rand real:\n");
print_type(myfunc(X));

real myfunc(rand real x) := avg(x+1);

print("Definitely real:\n");
print_type(myfunc(X));
print("Depends on promotion distances, currently real:\n");
print_type(myfunc(R));

print("\nTesting max() function promotions\n\n");

print("Should be int:\n");
print_type(max(3, 4, 5));
print("Should be real:\n");
print_type(max(3, 4.1, 5.0));
print("Should be ph int:\n");
print_type(max(3, 4, equilikely(5, 6)));
print("Should be ph real:\n");
print_type(max(0, expo(5.0)));
print("Should be ph real:\n");
print_type(max(erlang(3, 4.0), expo(5.0)));
print("Should be rand int:\n");
print_type(max(3, bernoulli(0.5), R));
print("Should be rand real:\n");
print_type(max(erlang(3, 4.0), expo(5.0), X));

print("\nTesting ord() promotions\n\n");

print("Should be int:\n");
print_type(ord(2, 1, 3, 5));
print("Should be real:\n");
print_type(ord(2, 1.1, 3.3, 5.5));
print("Should be ph int:\n");
print_type(ord(2, equilikely(1, 5), 3, 4));
print("Should be ph real:\n");
print_type(ord(2, erlang(1, 1.0), erlang(2, 2.0), erlang(3, 3.0)));
print("Should be rand int:\n");
print_type(ord(equilikely(1,3), 2, 4, 6));
print_type(ord(equilikely(1,3), 2, 4, R));
print("Should be rand real:\n");
print_type(ord(equilikely(1,3), 2.1, 4.2, 6.3));
print_type(ord(equilikely(1,3), 2, 4, X));


print("\nThe end\n");
