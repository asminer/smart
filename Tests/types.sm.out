
These should all have type void:

Expression print(null) has type: void
Expression print("foo\n") has type: void
Expression cond(true, print("bar\n"), null) has type: void

These should have type int:

Expression 2 has type: int
Expression a has type: int
Expression (2+a) has type: int
Expression (b-a) has type: int
Expression (a*b) has type: int

These should have type real:

Expression 5.1 has type: real
Expression (5.1+real(a)) has type: real
Expression (real(b)+c) has type: real
Expression (real(b)-c) has type: real
Expression (real(b)*c) has type: real
Expression (real(b)/c) has type: real
Expression (a/b) has type: real

Aggregate types:

Expression a:b:c has type: int:int:real
Expression 3:4:null has type: int:int:null

Testing function promotions:

Should be int:
Expression myfunc(3) has type: int
Should be rand int:
Expression myfunc(R) has type: rand int
Should be real:
Expression myfunc(3.5) has type: real
Should be rand real:
Expression myfunc(X) has type: rand real
Definitely real:
Expression myfunc(X) has type: real
Depends on promotion distances, currently real:
Expression myfunc(rand real(R)) has type: real

Testing max() function promotions

Should be int:
Expression max(3, 4, 5) has type: int
Should be real:
Expression max(real(3), 4.1, 5) has type: real
Should be ph int:
Expression max(3, 4, equilikely(5, 6)) has type: ph int
Should be ph real:
Expression max(real(0), ph real(expo(5))) has type: ph real
Should be ph real:
Expression max(erlang(3, 4), ph real(expo(5))) has type: ph real
Should be rand int:
Expression max(3, bernoulli(0.5), R) has type: rand int
Should be rand real:
Expression max(erlang(3, 4), rand real(expo(5)), X) has type: rand real

Testing ord() promotions

Should be int:
Expression ord(2, 1, 3, 5) has type: int
Should be real:
Expression ord(2, 1.1, 3.3, 5.5) has type: real
Should be ph int:
Expression ord(2, equilikely(1, 5), 3, 4) has type: ph int
Should be ph real:
Expression ord(2, erlang(1, 1), erlang(2, 2), erlang(3, 3)) has type: ph real
Should be rand int:
Expression ord(equilikely(1, 3), 2, 4, 6) has type: rand int
Expression ord(equilikely(1, 3), 2, 4, R) has type: rand int
Should be rand real:
Expression ord(equilikely(1, 3), 2.1, 4.2, 6.3) has type: rand real
Expression ord(equilikely(1, 3), real(2), real(4), X) has type: rand real

The end
